Extend parameter buffers to support Hantro G1 H.264

pH5 opened the pull request:
The Hantro G1 hardware H.264 decoder parses slices itself, with some assistance.
It requires the parsed `num_ref_idx_l[01]_default_active_minus1` and `idr_pic_id` syntax elements, as well as the size in bits of the `dec_ref_pic_marking()` syntax element and the total size in bits of the `pic_order_cnt_lsb`/`delta_pic_order_cnt_bottom`/`delta_pic_order_cnt[0]`/`delta_pic_order_cnt[1]` syntax elements.

These patches extend `VAPictureParameterBufferH264` and `VASliceParameterBufferH264` to allow supporting Hantro G1 decoders via [libva-v4l2-request](https://github.com/bootlin/libva-v4l2-request).

pH5 committed:
Add dec_ref_pic_marking() syntax element size to VASliceParameterBufferH264

The Hantro G1 hardware H.264 decoder requires this information.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

pH5 committed:
Add idr_pic_id syntax element to VASliceParameterBufferH264

The Hantro G1 hardware H.264 decoder requires this information.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

pH5 committed:
Add size of the pic_order_cnt related syntax elements to VASliceParameterBufferH264

Add the size in bits of the pic_order_cnt_lsb,
delta_pic_order_cnt_bottom, delta_pic_order_cnt[0], and
delta_pic_order_cnt[1] syntax elements.

The Hantro G1 hardware H.264 decoder requires this information.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

pH5 committed:
Add num_ref_idx_l[01]_default_active_minus1 syntax elements to VAPictureParameterBufferH264

The Hantro G1 hardware H.264 decoder requires this information.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

XinfengZhang commented on va/va.h:
```diff
@@ -3076,6 +3076,10 @@ typedef struct _VAPictureParameterBufferH264
+    uint8_t num_ref_idx_l1_default_active_minus1;
 
     /** \brief Reserved bytes for future use, must be zero */
     uint32_t                va_reserved[VA_PADDING_MEDIUM];
```
I understand that you add these field between the crevice of alignment, have you checked the size both on 64bit and 32 bit?

pH5 replied:
Yes, there is no difference between 64-bit (x86_64) and 32-bit (arm-v7a): In both cases, `sizeof(VAPictureParameterBufferH264)` is 672 bytes before and after the change. All member offsets stay the same.

`sizeof(VASliceParameterBufferH264)` is 3128 bytes before and after the change on both 64-bit and 32-bit. The only member offset that changes is `va_reserved`, which moves from 3112 to 3120.

ndufresne replied:
Shouldn't va_reseverd size be reduced to maintain overall structure size ?

pH5 replied:
Overall structure size is maintained by placing the new elements inside previously unused 32-bit alignment padding, see the comment above.

pH5 committed:
build: meson: bump api micro version

Bump the API version to 1.6.1 due to the fields added to
VAPictureParameterBufferH264 and VASliceParameterBufferH264.

CameronNemo commented:
Any word on whether this will be reviewed and/or merged?

CameronNemo commented:
@ndufresne thanks for the update. unfortunately firefox seems to depend on vaapi and cannot use v4l2 codecs directly. are there any non-libva options for firefox video acceleration?

okias commented:
> @ndufresne thanks for the update. unfortunately firefox seems to depend on vaapi and cannot use v4l2 codecs directly. are there any non-libva options for firefox video acceleration?

sadly, Firefox is in bad shape, so I assume Red Hat guys will not be contributing with another backend.

ndufresne commented:
Then feel free to carry on, you'll have to take the V4L2 Request VA driver maintenance though, as it no longer has a maintainer.

ndufresne commented on va/va.h:
```diff
@@ -3076,6 +3076,10 @@ typedef struct _VAPictureParameterBufferH264
+    /** \brief Same as the H.264 bitstream syntax element. */
+    uint8_t num_ref_idx_l0_default_active_minus1;
+    /** \brief Same as the H.264 bitstream syntax element. */
+    uint8_t num_ref_idx_l1_default_active_minus1;
```
Hantro/VSI hardware works at frame level. For my own understanding, how do one derive from slices this information ?

XinfengZhang replied:
from syntax level, num_ref_idx_lN_active_minus1 will be override by slice parameter (depend on whether num_ref_idx_active_override_flag is true). so I dont think that you just need a frame level syntax for it.  if there are multiple slice with different values. how to handle it?

ndufresne replied:
Ok, I'm not VSI employee, so I can only guess thing, let me know if that make sense. The HW will be parsing the value from the bitstream by itself, ignoring the values we have in VA-API structure per slice. Now, if it encounter a slice that does not have this syntax, it needs to use the default value, which we are adding here.

My question is, can we decuce from the information already in VA-API what is the default value ? Is it the maximum of all slices or something like this ?

XinfengZhang replied:
if application parse the bitstream, it always know whether the default value should be overridden.  it's why we just use slice parameter to set the value no matter it use default value or value from slice. it could cover all of them.  

so, I assume your problem is: application just parse part of the bitstream, SPS, PPS. then send other remaining bitstream to HW. HW may handle the sliceheader parser, and check whether it should be overridden. right?

for this case, the easiest way is sending a slice parameter, which just use to send these two parameters . other parameters in the slice parameter will be ignored.

ndufresne replied:
Said like this is sounds more like a hack, the application have no idea if the internal is slice or frame based. If you set the default in the firs slice, and the first slice wasn't using the defaults, you endup with the wrong value. I don't really know what this one is for and the semantic of it internally, that's a bit why I ask what to with only the squashed version (per slice final result). I think these two are somewhat legit request. But I guess it does not matter, since the driver would need reparse anyway to figure-out the skip bits parameters. I think the double parse overhead is unavoidable in the context of VA-API.with this type of HW.

ndufresne commented on va/va.h:
```diff
@@ -3149,9 +3155,18 @@ typedef struct _VASliceParameterBufferH264
     int16_t chroma_weight_l1[32][2];
     int16_t chroma_offset_l1[32][2];
+    /** \brief Size in bits of the dec_ref_pic_marking() syntax element. */
+    uint32_t dec_ref_pic_marking_bit_size;
```
The Hantro/VSI hardware do not have code to parse this, instead the size of that part is passed so the HW can simply skip it. This is hardware specific.

ndufresne commented on va/va.h:
```diff
@@ -3149,9 +3155,18 @@ typedef struct _VASliceParameterBufferH264
+     * Size in bits of the pic_order_cnt_lsb, delta_pic_order_cnt_bottom,
+     * delta_pic_order_cnt[0], and delta_pic_order_cnt[1] syntax elements.
+     */
+    uint32_t pic_order_cnt_bit_size;
```
Same, arguably, extending the API to support frame base or adding frame base hack might not fly. Perhaps users that want to support VA over V4L2 stateless uAPI (which is more flexible for HW vendor then VA) is to implement a bitstream parser in their VA driver. It's suboptimal, but you can then get all the parameters you need.
